Introduce `then` statements

hamishknight opened the pull request:
These allow multi-statement `if`/`switch` expression branches that can produce a value at the end by saying `then <expr>`. This is gated behind `-enable-experimental-feature ThenStatements` pending evolution discussion.

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibilty

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please SourceKit stress test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please SourceKit stress test

ahoppen commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -901,6 +908,99 @@ ParserResult<Stmt> Parser::parseStmtYield(SourceLoc tryLoc) {
+  // compatibility with the new SwiftSyntax parser, which will consider
+  // 'then . foo' to be a 'then' statement, not a binary operator.
+  if (peekToken().isAny(tok::l_paren, tok::l_square, tok::period)) {
+    auto nextLoc = Lexer::getLocForEndOfToken(SourceMgr, Tok.getLoc());
```
I’d name this variable `tokEndLoc`. `nextLoc` sounds like the location of `peekToken()`.

ahoppen reviewed (commented):
Looks good to me. Just a few minor comments.

hamishknight commented on include/swift/AST/Expr.h:
```diff
@@ -6144,6 +6145,16 @@ class SingleValueStmtExpr : public Expr {
 
+  /// Retrieves a resulting ThenStmt from the given BraceStmt, or \c nullptr if
+  /// the brace does not have a resulting ThenStmt.
+  static ThenStmt *getResultStmtFrom(BraceStmt *BS);
```
Yeah, I can't say I love this either. Renaming `ThenStmt` to `ResultStmt` doesn't seem right to me, as it seems way too generic of a name. I started using `ResultStmt` as a term in these methods for consistency with `getResultExprs`, but maybe it's fine not to have consistency with that.

hamishknight replied:
I renamed this

hamishknight commented on include/swift/Parse/Token.h:
```diff
@@ -108,7 +108,39 @@ class Token {
+      return true;
+
+    switch (Kind) {
+    case tok::period:
```
Yeah, that's because in SwiftSyntax, period can both be infix and postfix. Maybe we ought to at least add it to `OperatorLike` there. Or we could just remove it here as we don't actually rely on `tok::period` being matched here. I don't feel too strongly either way, any preference? A period does kind of seem "operator like".

hamishknight replied:
Oh hmm seems we rely on `OperatorLike` not matching period so that it's treated as unexpected for the `is not allowed in operator names` diagnostic in SwiftSyntax. Maybe we should just remove it here then

hamishknight replied:
Removed `tok::period` from here for now

hamishknight commented on lib/ASTGen/Sources/ASTGen/SourceFile.swift:
```diff
@@ -20,15 +23,34 @@ struct ExportedSourceFile {
 }
 
+extension Parser.ExperimentalFeatures {
+  static func from(_ context: BridgedASTContext?) -> Self {
```
Hmm yeah fair enough

hamishknight commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -901,6 +908,99 @@ ParserResult<Stmt> Parser::parseStmtYield(SourceLoc tryLoc) {
            status, YieldStmt::create(Context, yieldLoc, lpLoc, yields, rpLoc));
 }
 
+bool Parser::isContextualThenKeyword(bool preferExpr) {
```
Hmm, but then it wouldn't be consistent with all the other `isContextual*` functions in the C++ parser, I guess we could rename all of those, but I think I would rather just keep them as-is given this is soon to become legacy code

ahoppen replied:
Ah, I didn’t know there was a `isContextual*Keyword` family in the C++ parser. Fine to keep it with that in mind.

hamishknight commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -901,6 +908,99 @@ ParserResult<Stmt> Parser::parseStmtYield(SourceLoc tryLoc) {
+
+  auto exprLoc = Tok.getLoc();
+
+  // Issue a warning when the expression is on a different line than
+  // the 'then' keyword, but both have the same indentation.
+  if (SourceMgr.getLineAndColumnInBuffer(thenLoc).second ==
+      SourceMgr.getLineAndColumnInBuffer(exprLoc).second) {
+    diagnose(exprLoc, diag::unindented_code_after_then);
+    diagnose(exprLoc, diag::indent_expression_to_silence);
+  }
```
Depending on where we land with the newline rule, we may not even need this logic in the first place

hamishknight commented on include/swift/AST/Stmt.h:
```diff
@@ -304,6 +304,38 @@ class YieldStmt final
+
+  /// Create an implicit ThenStmt.
+  ///
+  /// Note that such statements will be elided during the result builder
+  /// transform.
```
Added test for explicit `then` statement

ahoppen replied:
Ah, that makes sense. Thanks for clarifying.

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

gottesmm commented:
@hamishknight I am doing something similar that is going to need to be gated by an experimental feature in SwiftSyntax. Is it possible for you to split out the addition of the experimental feature support into a separate commit in this same PR. This will make it easier for people to cargo cult this work forward by just looking at a small commit that implements this rather than the large commit here. (I am not just doing this for myself, I am thinking down the line about other people).

gottesmm commented:
Or actually from talking @ahoppen this is going to take some time for you to do. I am going to see if I can cargo cult some of your work around the experimental feature since I need the same thing for reference bindings.

hamishknight commented:
@gottesmm Hey Michael, I already did this :) See https://github.com/apple/swift/pull/67826 + https://github.com/apple/swift-syntax/pull/2041. There's one more commit on the swift-syntax side that's needed to enable testing of experimental features (https://github.com/apple/swift-syntax/pull/2000/commits/b0414b188918ca2a61413f0e5a21ae83a1be9b1d), I can land that separately if needed.

gottesmm commented:
@hamishknight can you land it separately in that case?

hamishknight commented:
Sure thing: https://github.com/apple/swift-syntax/pull/2051

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

airspeedswift commented:
apple/swift-syntax#2000

@swift-ci please build toolchain

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

xedin commented on test/stmt/then_stmt_disabled.swift:
```diff
@@ -0,0 +1,11 @@
+let x = if .random() {
+  print("hello")
+  then 0
+  // expected-error@-1 {{cannot find 'then' in scope}}
```
Should this be diagnosed as a disabled feature instead if `then` here otherwise would be valid?

hamishknight replied:
This is a source breaking change, so we can't generally diagnose it as a disabled feature, we could apply a heuristic to generate that diagnostic, but I'm not convinced it's worth it.

xedin replied:
Maybe it's not worth it if we are going to enable this by default soon, but it would be great for user experience to hint that all they need is to enable a flag in situations when it's syntactically unambiguous that they meant `then` statement.

xedin commented on include/swift/Sema/ConstraintLocator.h:
```diff
@@ -103,8 +103,14 @@ enum class ConversionRestrictionKind;
+  Implicit,
+
+  /// Implicitly written '<expr>' in a single expr closure body.
+  ImplicitInInSingleExprClosure
```
One `In` too many? :)

hamishknight replied:
Good catch!

xedin commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -2159,6 +2175,38 @@ class SyntacticElementSolutionApplication
+    // Note we're defensive here over whether we're in a SingleValueStmtExpr,
+    // as we don't diagnose out-of-place ThenStmts until MiscDiagnostics. If we
+    // have an out-of-place ThenStmt, just rewrite the expr without a contextual
+    // type.
```
Is there not enough information to determine whether a `then` is out of place during constraint generation for a syntactic element? We could fail invalid uses in regular solving and produce a fix in diagnostic mode?

hamishknight replied:
I initially wanted to keep all the logic together in the MiscDiagnostics walker, but yeah I think doing it this way is probably better, updated to use a fix.

xedin replied:
Thank you!

hamishknight committed:
Introduce `then` statements

These allow multi-statement `if`/`switch` expression
branches that can produce a value at the end by
saying `then <expr>`. This is gated behind
`-enable-experimental-feature ThenStatements`
pending evolution discussion.

hamishknight committed:
Move out-of-place ThenStmt checking to constraint generation

hamishknight commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please test

airspeedswift commented:
https://github.com/apple/swift-syntax/pull/2000

@swift-ci please build toolchain
